network/p2p/dht_test.go

```diff
@@ -0,0 +1,219 @@
+package p2p
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	golog "github.com/ipfs/go-log"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	discovery "github.com/libp2p/go-libp2p-discovery"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/metrics"
+	flownet "github.com/onflow/flow-go/network"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+type DHTTestSuite struct {
+	suite.Suite
+	ctx    context.Context
+	cancel context.CancelFunc // used to cancel the context
+}
+
+// TestDHTTestSuite test the libp2p pubsub with DHT for discovery
+func TestDHTTestSuite(t *testing.T) {
+	suite.Run(t, new(DHTTestSuite))
+}
+
+// SetupTests initiates the test setups prior to each test
+func (suite *DHTTestSuite) SetupTest() {
+	suite.ctx, suite.cancel = context.WithCancel(context.Background())
+}
+
+// TestPubSub checks if nodes can subscribe to a topic and send and receive a message on that topic. The DHT discovery
+// mechanism is used for nodes to find each other.
+func (suite *DHTTestSuite) TestPubSubWithDHTDiscovery() {
+	defer suite.cancel()
+	topic := flownet.Topic("/flow/" + unittest.IdentifierFixture().String())
+	count := 5
+	golog.SetAllLoggers(golog.LevelFatal) // change this to Debug if libp2p logs are needed
+
+	// Step 1: Creates nodes
+	// Nodes will be connected in a hub and spoke configuration where one node will act as the DHT server,
+	// while the other nodes will act as the client.
+	// The hub-spoke configuration should eventually converge to a fully connected graph as all nodes discover
+	// each other via the central node.
+	// We have less than 6 nodes in play, hence the full mesh. LibP2P would limit max connections to 12 if there were
+	// more nodes.
```

Comment:
Yup - so this libp2p pubsub will try and maintain full mesh peers between [Dlo and Dhi number of nodes](https://github.com/libp2p/go-libp2p-pubsub/blob/master/gossipsub.go#L34-L35)
https://docs.libp2p.io/concepts/publish-subscribe/#full-message.

If the number of nodes are less than Dlo then it will connect to all those peers.

Reply:
In this case, even if we have 6 nodes in total, we will still have a full mesh right? Because each node will have a connection to 5 other nodes, and the default Dlo is 5?

Reply:
yes, here we can expect a full mesh to be always formed. If the count was bumped to a value greater than 5 then libp2p will most likely still connect to all the other nodes initially and then start dropping connections later however  the test may fail once in a while.

Reply:
Can you output the routing table contents for each of the nodes? https://pkg.go.dev/github.com/libp2p/go-libp2p-kad-dht#IpfsDHT.RoutingTable

I'm curious to see what it looks like

Reply:
sure, I will do that.

Reply:
Example of the routing table:
Routing Table, bs = 20, Max latency = 60000000000
	bucket: 0
		- 16Uiu2HAmTxufgKViWfranSwigvcEFQUoheEFphF2sFQkJuu7j7fS 260.37µs

Reply:
that output is from the unstaked node, the staked node printing an empty routing table.

Reply:
That's strange, I wonder if it's because they are all running in client mode.

Did you print the routing table after the full mesh was formed? or before?

Reply:
printed it after the full mess was formed ( the full connected test passed) and then I dumped the table